### PR TITLE
add NewStack(), which generates a stack trace

### PR DESCRIPTION
**Pull request by gregwebs**
This avoids using hacky code to get a trace.

### CODE REVIEW

**Comment by gregwebs** [comment]
@lysu @coocood

**Comment by coocood** [comment]
Where this function will be used except for the unexported `stackTracke`?
To replace `util.GetStack` ?

**Comment by gregwebs** [comment]
@coocood I am not understanding your references. `StackTracer` is now exported. Where is `util.GetStack`?
I wanted to use this function here: https://github.com/pingcap/pd/pull/1200/files#diff-f10e5c1ee6d26ee1746a0a9188a67d4dR76

**Comment by lysu** [comment]
@coocood  stackTrace is an test method, I think it should be a util method, maybe someone want re-encapsulate errors to biz errors api need this by control `position` value??(tidb util.GetStack can not do this)

but I have same question why we should use `NewStack` instead of `GetStack`

**Comment by lysu** [comment]
@gregwebs  `util.GetStack` is https://github.com/pingcap/tidb/blob/99aba76c90afb05bef336b51aff353e874d9e178/util/misc.go#L49

**Comment by gregwebs** [comment]
@coocood @lysu okay, I see you are referencing a function in the tidb repo. Yes, they are almost the same, but the types are slightly different. Mostly this is just exposing the function from this library (with the types and interfaces of this library). It does allow removing some of the top callers from the stack when building on it.

**Comment by lysu** [comment]
I think this function is look good to me, previous I meet same requirement too~

File: stack.go [hunk 1]
```diff
@@ -151,3 +151,21 @@ func funcname(name string) string {
...
+
+// NewStack is for library implementers that want to generate a stack trace.
+// Normally you should use AddStack to get an error with a stack trace.
+// Stack trace information can be generated by calling this function.
+// It can then be turned into a stack trace by calling .StackTrace()
+//
+// This function takes an optional argument for the position in the stack to start from.
+// This avoids putting stack generation function calls in the stack trace.
+// It defaults to 1, which removes the NewStack function from the stack trace.
+func NewStack(position ...int) StackTracer {
```
Comment (coocood) [inline stack.go]: I think variable args should be used for multiple values, not for optional.
How about that we make `position` a fixed argument, then add 1 internally?
The default usage is `NewStack(0)`.

Comment (gregwebs) [reply stack.go]: I agree its not great, but that's golang :(. I think `NewStack(int)` is possible, but we cannot have any internal addition. In that case I would just consider exposing two functions, one that uses just 1.

Comment (gregwebs) [reply stack.go]: @coocood Okay, I am requiring the argument now. But there is no internal addition (beyond what was already used in callers()), so normal usage will be `NewStack(1)`.

**Comment by gregwebs** [comment]
Comments addressed. I just changed the settings to require an approval before merge.

**Comment by coocood** [comment]
@gregwebs 
Why the skip value has to be started from 1?

**Comment by gregwebs** [comment]
@coocood you have the option to start it at zero. That is done in the test case. Maybe that is not useful outside of the test case though?

**Comment by coocood** [comment]
Then we can modify the test case.

**Comment by coocood** [comment]
Or pass `-1` in the test case?

**Comment by gregwebs** [comment]
I think `0` is useful and makes sense: that is the line where the stack trace is generated. `NewStack` itself does not show up.

**Comment by gregwebs** [comment]
I updated the docs for the skip parameter, which were incorrect, hopefully this makes sense now.

**Comment by coocood** [comment]
LGTM

@lysu PTAL

**Pull request merged**